### DISCUSSION TITLE
R-UDF , TSA entry points but with access to full spectral profile

### DISCUSSION THREAD

**Discussion by JoLeBelge** [davidfrantz/force#365] [Q&A]
Dear FORCE community, 
thank you for your open source work and documentation. greatly appreciated.

I discovered FORCE a few weeks ago and its help me a lot for the processing of long time serie of S2 (complete ts from 2017 to 2025 70% cloud free coverage for south of Belgium). I have exported the 8 years TSI results from FORCE to SITS (R stat) package to model the forest dieback in oaks stands with differents algorithms, in particular temporal CNN. My models use all dates of the TSI and multiple vegetation index. Until now I have worked on a small number of datacube (of 5kmx5km), the one that cover my training data, so processing speed wasn't a concern. Now I want to use my trained model (sits model , which is a torch model wrapped in R) to compute predictions for the whole country - with a forest mask of course.

I believe FORCE is very adequate, so I took a closer look to the R-UDF functioning in order to run my sits model on the TSI. But if I understand well, the TSI entry point for R-UDF does not give access to multiple bands at once. So my question, is there a way to access multiple "index" from the TSI results and to perform processing in R program?

thanks you in advance,
Jonathan Lisein - forestry researcher

**Comment by davidfrantz** [comment]
Hi Jonathan,

welcome here! 

It may require a small workaround, but yes, this is possible. The trick is to use the ARD entry point, i.e., the ``UDF`` submodule in ``force-higher-level``.

You can use the Level 2 ARD directly, which provides access to all spectral bands simultaneously. The disadvantage is that you would need to perform the index calculation in the UDF (though this is relatively easy), as well as the interpolation (which is probably a bit more challenging).

The alternative is to use the precomputed TSI and trick FORCE into thinking they are Level 2 ARD. This requires some renaming and reorganisation of the data. You would need to create layerstacks per date, which would need to be named according to the Level 2 naming pattern. The stacking and renaming can be done using the vrt format; thus, you do not need to copy the data physically. Important is, that the number of bands in your index-stack corresponds to the number of bands that are expected of the sensor (i.e., the sensor string in the file name). As an example, you may use LND08 or LNDLG if you have 6 indices. SEN2A or SEN2L for 10 indices. sen2a for 4 indices. There are a couple more implemented. If you share how many indices you have, I can make a recommendation on the "fake" sensor name.

Hope this helps,
David

**Comment by JoLeBelge** [comment]
Hi David,
Thank you for your quick answer!
I am digging for solution number 2; tricking FORCE to load level 3 products as if they were level 2 ARD. I have created vrt files with 10 bands (for SEN2A) and I have set the description of the band according to FORCE conventions. It seems to work until now. But it is a bit a mess in all my product's directories :-)

Jonathan

**Reply by davidfrantz** [reply]
true ;)

You could also generate the VRTs in a new cube - in case the mess bothers you
